### PR TITLE
Stratified cross-validation

### PR DESCRIPTION
**Pull request by mmerce**

### CODE REVIEW

File: cross-validation/boosted-ensemble/script.whizzml [hunk 1]
```diff
@@ -172,6 +195,29 @@
...
+                                                        "(= "
+                                                        "(f {objective-id})"
+                                                        " {{cat}})")
+                                        "name" (str (dataset "name")
+                                                    " - "
+                                                    cat)})
+                       k-folds))
+        ds-ids (for (index (range 0 (count (cat-ds-ids 0))))
+                 (create-dataset {"origin_datasets" (map (lambda (x) (x index))
+                                                         cat-ds-ids)})))
```
Comment (jaor) [inline cross-validation/boosted-ensemble/script.whizzml]: if i'm reading this correctly, you want to map over the transposed
`cats-ds-ids`, right? if so, a shorter way of transposing `m`
if `(apply map list m)`

```
ds-ids (map (lambda (ids) (create-dataset {"origin_datasets" ids}))
            (apply map list cats-ds-ids))
```

we could have a `transpose` defined in he stdlib... if you use this
trick here, i would define a little `_transpose` function to make
it clearer.

File: cross-validation/deepnet/script.whizzml [hunk 2]
```diff
@@ -126,6 +129,23 @@
...
 
 
+;; check-minority-instances
+;;
+;; Checks whether the number of instances in the minority class is greater
+;; than the number of folds
+(define (check-minority-instances k-folds dataset-id objective-id)
+  (let (dataset (fetch dataset-id)
+        min-cat (last (dataset ["fields" objective-id "summary" "categories"]
+                               []))
```
Comment (jaor) [inline cross-validation/deepnet/script.whizzml]: this is going to throw an exception if there are no categories... did you mean:

```
(last (dataset ["fields" ...] []) [])
```

or

```
(last (datasest ["fields" ...] [[]])
```

File: cross-validation/deepnet/script.whizzml [hunk 3]
```diff
@@ -126,6 +129,23 @@
...
+        min-inst (if (empty? min-cat)
+                     0
+                     (min-cat 1)))
```
Comment (jaor) [inline cross-validation/deepnet/script.whizzml]: `(min-cat 1 0)` does the same thing...

File: cross-validation/deepnet/script.whizzml [hunk 4]
```diff
@@ -159,6 +179,30 @@
     (fields [objective-id "name"] false)))
 
 
+;; Stratified sampling: Keeping proportions in sampling across classes
+
+(define (stratified-k-folds dataset k-folds objective-id)
+  (let (summary (dataset ["fields" objective-id "summary" "categories"] [])
+        cat-names (map (lambda (x) (x 0)) summary)
```
Comment (jaor) [inline cross-validation/deepnet/script.whizzml]: nit, but `(map head summary)` is shorter

File: cross-validation/deepnet/script.whizzml [hunk 5]
```diff
@@ -159,6 +179,30 @@
...
+                                                        "(= "
+                                                        "(f {objective-id})"
+                                                        " {{cat}})")
+                                        "name" (str (dataset "name")
+                                                    " - "
+                                                    cat)})
+                       k-folds))
+        ds-ids (for (index (range 0 (count (cat-ds-ids 0))))
+                 (create-dataset {"origin_datasets" (map (lambda (x) (x index))
+                                                         cat-ds-ids)})))
```
Comment (jaor) [inline cross-validation/deepnet/script.whizzml]: if we define a `_transpose` we can reuse it here

File: cross-validation/ensemble/script.whizzml [hunk 6]
```diff
@@ -124,6 +131,22 @@
...
+;;
+;; Checks whether the number of instances in the minority class is greater
+;; than the number of folds
+(define (check-minority-instances k-folds dataset-id objective-id)
+  (let (dataset (fetch dataset-id)
+        min-cat (last (dataset ["fields" objective-id "summary" "categories"]
+                               []))
+        min-inst (if (empty? min-cat)
+                     0
+                     (min-cat 1)))
```
Comment (jaor) [inline cross-validation/ensemble/script.whizzml]: same as before

File: cross-validation/ensemble/script.whizzml [hunk 7]
```diff
@@ -171,6 +194,28 @@
...
+                                                        "(= "
+                                                        "(f {objective-id})"
+                                                        " {{cat}})")
+                                        "name" (str (dataset "name")
+                                                    " - "
+                                                    cat)})
+                       k-folds))
+        ds-ids (for (index (range 0 (count (cat-ds-ids 0))))
+                 (create-dataset {"origin_datasets" (map (lambda (x) (x index))
+                                                         cat-ds-ids)})))
```
Comment (jaor) [inline cross-validation/ensemble/script.whizzml]: same as before

File: cross-validation/ensemble/script.whizzml [hunk 8]
```diff
@@ -171,6 +194,28 @@
 (define (get-objective-name dataset objective-id)
   (dataset ["fields" objective-id "name"] false))
 
+;; Stratified sampling: Keeping proportions in sampling across classes
+
+(define (stratified-k-folds dataset k-folds objective-id)
+  (let (summary (dataset ["fields" objective-id "summary" "categories"] [])
+        cat-names (map (lambda (x) (x 0)) summary)
```
Comment (jaor) [inline cross-validation/ensemble/script.whizzml]: same as before

File: cross-validation/ensemble/script.whizzml (hunk 7 above)
Comment (jaor) [reply cross-validation/ensemble/script.whizzml]: i'm stopping now :) perhaps we should consider a library?

**Review by jaor — CHANGES_REQUESTED** [review]
a few nits... this is hard to maintain (and review) with so much code repetition, not sure if we could use a library here...

**Comment by mmerce** [comment]
The cross-validation library has been separated and your suggestions added.

**Review by jaor — APPROVED** [review]

**Pull request merged**